### PR TITLE
Cypress tech memo, re: test-e2e-deployment

### PR DESCRIPTION
**Pull request by andrew-jameson**
## Summary of Changes
_Provide a brief summary of changes_
Pull request relates to #3141 
Additionally, has some minor developer quality of life improvements:
1. Task to run cypress
2. Reduce CircleCI workflows kicked off by a commit to existing PR

## How to Test
N/A

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [ ] N/A

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [ ] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [ ] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [ ] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [ ] Are backend code style checks passing on CircleCI?
+ [ ] Are frontend code style checks passing on CircleCI?
+ [ ] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [ ] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [ ] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [ ] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [ ] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

File: .github/workflows/build-pr.yml [hunk 1]
```diff
@@ -27,6 +27,7 @@ on:
       - main
       - master
       - 'release/**'
+      - 'bug/3141-e2e-minimal'
```
Comment (andrew-jameson) [inline .github/workflows/build-pr.yml]:
```suggestion
```

File: .circleci/deployment/workflows.yml [hunk 2]
```diff
@@ -121,9 +127,11 @@
             branches:
               only:
                 - develop
+                - bug/e2e-minimal
```
Comment (andrew-jameson) [inline .circleci/deployment/workflows.yml]:
```suggestion
```

File: .circleci/deployment/workflows.yml [hunk 3]
```diff
@@ -93,13 +97,15 @@
...
                 - master
+                - bug/e2e-minimal
       - deploy-develop:
           requires:
             - deploy-infrastructure-staging
           filters:
             branches:
               only:
                 - develop
+                - bug/e2e-minimal
```
Comment (andrew-jameson) [inline .circleci/deployment/workflows.yml]:
```suggestion
```

File: .circleci/deployment/workflows.yml [hunk 4]
```diff
@@ -93,13 +97,15 @@
                 - develop
                 - main
                 - master
+                - bug/e2e-minimal
```
Comment (andrew-jameson) [inline .circleci/deployment/workflows.yml]:
```suggestion
```

File: .circleci/deployment/workflows.yml [hunk 5]
```diff
@@ -41,6 +44,7 @@
             branches:
               ignore:
                 - develop
+                - bug/e2e-minimal
```
Comment (andrew-jameson) [inline .circleci/deployment/workflows.yml]:
```suggestion
```

File: .circleci/deployment/workflows.yml [hunk 6]
```diff
@@ -6,6 +6,7 @@
             - equal: [ master, << pipeline.git.branch >> ]
             - equal: [ main, << pipeline.git.branch >> ]
             - equal: [ develop, << pipeline.git.branch >> ]
+            - equal: [ bug/e2e-minimal, << pipeline.git.branch >> ]
```
Comment (andrew-jameson) [inline .circleci/deployment/workflows.yml]:
```suggestion
```

File: .circleci/deployment/workflows.yml [hunk 7]
```diff
@@ -21,13 +22,15 @@
             branches:
               ignore:
                 - develop
+                - bug/e2e-minimal
```
Comment (andrew-jameson) [inline .circleci/deployment/workflows.yml]:
```suggestion
```

File: .circleci/deployment/workflows.yml [hunk 8]
```diff
@@ -21,13 +22,15 @@
...
                 - main
                 - master
       - deploy-infrastructure-staging:
           filters:
             branches:
               only:
                 - develop
+                - bug/e2e-minimal
```
Comment (andrew-jameson) [inline .circleci/deployment/workflows.yml]:
```suggestion
```

File: .circleci/deployment/workflows.yml [hunk 9]
```diff
@@ -121,9 +127,11 @@
...
       - make_erd: # from ../util folder
           filters:
             branches:
               only:
                 - develop
                 - master
+
```
Comment (andrew-jameson) [inline .circleci/deployment/workflows.yml]:
```suggestion
```

File: tdrs-frontend/cypress/e2e/common-steps/common-steps.js [hunk 10]
```diff
@@ -6,6 +6,10 @@ When('{string} visits the home page', (username) => {
   cy.contains('Sign into TANF Data Portal', { timeout: 30000 })
 })
 
+When('{string} visits', (username) => {
+  cy.visit('/')
+})
+
```
Comment (andrew-jameson) [inline tdrs-frontend/cypress/e2e/common-steps/common-steps.js]:
```suggestion
```

File: tdrs-frontend/cypress/e2e/accounts/accounts.js [hunk 11]
```diff
@@ -81,6 +81,27 @@ Then('{string} cannot log in', (username) => {
...
+
+  // cy.authcheck(userSessionId, userCsrfToken)
+  // cy.visit('/').then(() => {
+  //  cy.contains('Inactive Account').should('exist')
+  // })
+  cy.wait(30000).then(() => {
+   cy.contains('Inactive Account').should('exist')
+  })
+})
+
```
Comment (andrew-jameson) [inline tdrs-frontend/cypress/e2e/accounts/accounts.js]:
```suggestion
```

File: tdrs-frontend/cypress/e2e/accounts/accounts.feature [hunk 12]
```diff
@@ -39,3 +39,4 @@ Feature: Users can create and manage their accounts
         Then 'new-cypress@teamraft.com' can see the hompage
         When The admin sets the approval status of 'new-cypress@teamraft.com' to 'Deactivated'
         Then 'new-cypress@teamraft.com' cannot log in
+
```
Comment (andrew-jameson) [inline tdrs-frontend/cypress/e2e/accounts/accounts.feature]:
```suggestion
```

File: docs/Technical-Documentation/tech-memos/cypress-auth.md [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+* Significant changes to backend authentication
+* New Cypress workflows beyond our end-to-end test against deployed develop branch
+
+## Method/Design
+
+### Abstracted Gherkin Steps
+Presently, many of the defined Javascript functions for a given Gherkin step are bespoke or single-use instead of abstracted and should be adapted. Additionally, it was found that sessions were lingering between Gherkin scenarios as we did not have generic `setup` and `teardown` implementations ahead of these. Sufficient utilization of abstraction within the scenarios which are now doing setup/teardown between scenarios and proper session management should result in a cleaner Cypress execution and make future additions simpler.
+
+### Abstracted utility authentication functions
+Our current cypress implementation has writes Gherkin scenarios `accounts.feature` which relies on definitions in `accounts.js`, `common-steps.js`, and finally `commands.js` which handle authentication in different ways for different scenarios (e.g., `login()`, `adminLogin()`, and `adminApiRequest()`)
```
Comment (elipe17) [inline docs/Technical-Documentation/tech-memos/cypress-auth.md]: Small grammar error here: `current cypress implementation has writes Gherkin`.

File: docs/Technical-Documentation/tech-memos/cypress-auth.md [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+
+## Out of Scope
+* Any changes to frontend ReactJS and Nginx apps
+* Significant changes to backend authentication
+* New Cypress workflows beyond our end-to-end test against deployed develop branch
+
+## Method/Design
+
+### Abstracted Gherkin Steps
+Presently, many of the defined Javascript functions for a given Gherkin step are bespoke or single-use instead of abstracted and should be adapted. Additionally, it was found that sessions were lingering between Gherkin scenarios as we did not have generic `setup` and `teardown` implementations ahead of these. Sufficient utilization of abstraction within the scenarios which are now doing setup/teardown between scenarios and proper session management should result in a cleaner Cypress execution and make future additions simpler.
```
Comment (elipe17) [inline docs/Technical-Documentation/tech-memos/cypress-auth.md:26]: Could you provide an example of what level of abstraction your are driving towards?

Comment (elipe17) [reply docs/Technical-Documentation/tech-memos/cypress-auth.md:26]: Or is that what the `### Abstracted utility authentication functions` section is covering?

Comment (andrew-jameson) [reply docs/Technical-Documentation/tech-memos/cypress-auth.md:26]: The abstracted section below is for the JavaScript side. I will revise this section to be more verbose on how I imagine we might restructure the Gherkin. Reading through it, especially without Jan's clearing of cookies, one scenario needed to be run before another, etc. Assuming we get the setup/teardown, these scenarios can be independent. I'll include psuedo-code examples as we discussed Friday I think on how tech memos ought to look.

File: docs/Technical-Documentation/tech-memos/cypress-auth.md (hunk 13 above)
Comment (andrew-jameson) [reply docs/Technical-Documentation/tech-memos/cypress-auth.md]: Thank you, fixing!

Comment (andrew-jameson) [inline docs/Technical-Documentation/tech-memos/cypress-auth.md]:
```suggestion
Our current Cypress implementation has Gherkin scenarios `accounts.feature` which relies on definitions in `accounts.js`, `common-steps.js`, and finally `commands.js` which handle authentication in different ways for different scenarios (e.g., `login()`, `adminLogin()`, and `adminApiRequest()`)
```

File: docs/Technical-Documentation/tech-memos/cypress-auth.md (hunk 14 above)
Comment (jtimpe) [reply docs/Technical-Documentation/tech-memos/cypress-auth.md:26]: we might need to update cypress and/or cypress-cucumber-preprocessor to get these (i couldn't get them to work in the 5 minutes i spent messing with it), but we should have `BeforeAll` and `BeforeEach` hooks available

https://github.com/badeball/cypress-cucumber-preprocessor/blob/master/docs/cucumber-basics.md#hooks

**Comment by lhuxraft** [comment]
12/4: Andrew to re-request reviews

**Review by elipe17 — APPROVED** [review]
Looking good. Thanks for adding the extra steps!

**Review by jtimpe — APPROVED** [review]

**Review by raftmsohani — APPROVED** [review]
LGTM!

**Pull request merged**